Fixes #31947 - add grub2/ks validation to jenkins rake task

lzap opened the pull request:
Adds ksvalidator and grub2 syntax checkers for selected template snapshots.

Currently only RHEL7 profile passes, for RHEL8 we need to fix authconfig which was deprecated and ksvalidator already fails. I added a task but it's not called from the "all" task.

I've found that ptable mock had some errors, I've fixed this as well to make ksvalidator passing.

The patch blindly assumes that both tools are on PATH on our Jenkins nodes. I would like someone from the infra team to help me with setting up the environment. We need latest version of pykickstart installed, for grub2 any version will do as the parsing validation is only shallow.

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

lzap committed:
Fixes #31947 - add grub2/ks validation to jenkins rake task

ekohl commented on lib/tasks/test.rake:
```diff
@@ -16,4 +16,36 @@ namespace :test do
+    end
+
+    desc "Validate selected template snapshots"
+    task :all => [:kickstart_el7, :grub2]
```
What about kickstart_el8? Why don't we have a `kickstart` task that goes over all supported RHEL versions? I'm sure we'll have a RHEL9 at some point in the future.

lzap replied:
As I said in the description, it currently fails due to deprecation which is already reported as an error by the tool. I don't want to sove this in this PR, thus I am creating the task but we would not call this from jenkins for now. This is security related and it will involve some more testing. FYI it's realated to this: https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html-single/configuring_authentication_and_authorization_in_rhel/index#converting-your-scripts-from-authconfig-to-authselect_configuring-user-authentication-using-authselect

ezr-ondrej reviewed (commented):
I really love the syntax checker addition!
But shouldn't we run it separatelly and only add that as a step in jenkins?

ezr-ondrej commented:
@lzap?

lzap commented:
> But shouldn't we run it separatelly and only add that as a step in jenkins?

I do not understand.

ezr-ondrej commented:
Well I don't think we should add this as part of `rake jenkins:unit`, but rather have e.g. `rake jenkins:syntax` and add it to jenkins manually.

lzap commented:
Okay `jenkins:syntax` it is.

ezr-ondrej commented on lib/tasks/test.rake:
```diff
@@ -16,4 +16,36 @@ namespace :test do
+    task :kickstart_el7 do
+      Dir.glob('test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/*Kickstart*').each do |file|
+        puts "Validating #{file}"
+        `ksvalidator --version RHEL7 "#{file}"`
+        abort("invalid kickstart: #{file}") unless $?.success?
```
Why did you choose backtits over `system` here, when you don't work with the output?
`system` would give you the result without touching global variables:

```suggestion
        result = system("ksvalidator --version RHEL7 '#{file}'")
        abort("invalid kickstart: #{file}") unless result
```

lzap replied:
Well `$?` is thread local, but indeed looks nicer.

ezr-ondrej commented on lib/tasks/test.rake:
```diff
@@ -16,4 +16,36 @@ namespace :test do
+    task :kickstart_el8 do
+      Dir.glob('test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/*Kickstart*').each do |file|
+        puts "Validating #{file}"
+        `ksvalidator --version RHEL8 "#{file}"`
```
Given EL7 and EL8 differs just by this one parameter, would it be worth to have common kickstart taks, that would accept the EL version param and these two would just `invoke` that task with correct version parameter?

lzap replied:
I have no idea how to execute tests for both el7 and el8 if you write just `rake syntax:kickstart` tho.

ezr-ondrej commented on test/factories/host_related.rb:
```diff
@@ -29,7 +29,7 @@ def set_environment_taxonomies(host_or_hostgroup, environment = host_or_hostgrou
   factory :ptable do
     sequence(:name) { |n| "ptable#{n}" }
-    layout { 'zerombr\nclearpart --all    --initlabel\npart /boot --fstype ext3 --size=<%= 10 * 10 %> --asprimary\npart /     --f   stype ext3 --size=1024 --grow\npart swap  --recommended' }
+    layout { "zerombr\nclearpart --all    --initlabel\npart /boot --fstype ext3 --size=<%= 10 * 10 %> --asprimary\npart /     --fstype ext3 --size=1024 --grow\npart swap  --recommended" }
```
Nit: I'd like this in separate commit bit better, but it's just me nitpicking...

lzap replied:
I have to correct it here because kickstart would fail to validate :)

ezr-ondrej reviewed (commented):
I really love this, we have not much validation over our templates and anything we can add will be great! Commenting just on some nits.

ekohl reviewed (commented):
Why don't you implement it as tests? We already have functional tests, feels like it could be another category like that. The benefit is that you get actual JUnit test output rendering rather than having to dive into the console output.

Just like we have here:
https://github.com/theforeman/foreman/blob/5c3d913c14746bae84d90bdfd8c9fb11d3cf4017/lib/tasks/test.rake#L4-L10

This also follows a pattern that's in Rails itself:
https://github.com/rails/rails/blob/main/railties/lib/rails/test_unit/testing.rake

lzap commented:
You are right, I wanted to avoid loading Rails environment but probably it is not big deal.

lzap commented:
So I introduced new test type `test/external` and `test:external` which requires external tools to be present. If a tool is not available, test is skipped. I also added `jenkins:external` which will allow us to configure where to call these tests.

I had to remove EL8 kickstart testing tho, because I just realized that snapshots are all EL7 only, so this cannot be workarounded in the ERB template. We need to upgrade snapshot tests to EL8 someday, not on the table for today (I have a huge snaphot refactoring in the other PR which I want to finish first).

ekohl commented on test/external/kickstart_syntax_test.rb:
```diff
@@ -0,0 +1,23 @@
+
+class KickstartSyntaxTest < ActiveSupport::TestCase
+  # Kickstart snapshots are generated only for EL7
+  ["RHEL7"].each do |version|
+    Dir.glob('test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/*Kickstart*').each do |file|
+      test "#{version} for #{file}" do
+        ksvalidator(file, version)
+      end
+    end
+  end
```
I think you can use `describe` as well:
```suggestion
  ["RHEL7"].each do |version|
    describe version do
      Dir.glob('test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/*Kickstart*').each do |file|
        test file do
          ksvalidator(file, version)
        end
      end
    end
  end
```

This may show up cleaner in the reporting but not 100% sure. I have more experience with rspec than minitest.

lzap replied:
Yeah its called `context` in minitest, I can do that sure.

ekohl commented on test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/XenServer default answerfile.snap.txt:
```diff
@@ -1,6 +1,10 @@
 <?xml version="1.0"?>
 <installation mode="fresh" srtype="lvm">
-zerombr\nclearpart --all    --initlabel\npart /boot --fstype ext3 --size=100 --asprimary\npart /     --f   stype ext3 --size=1024 --grow\npart swap  --recommended  <keymap>us</keymap>
+zerombr
+clearpart --all    --initlabel
+part /boot --fstype ext3 --size=100 --asprimary
+part /     --fstype ext3 --size=1024 --grow
+part swap  --recommended  <keymap>us</keymap>
```
Just reading this, I highly doubt this is correct. Doesn't look like valid XML. Maybe something for another day.

It does look like we need the correct OS to be set for this.

lzap replied:
I am fixing the newline, but definitely this does not look good. This is probably broken for some time I guess, I do not use Xen and I am not planning to :)

lzap replied:
Ah I think I know. This is an incorrect snapshot test, I am removing it from `snapshots.yaml` because in snapshots we only test with EL7 and this obviously needs a different partitioning template.

There is a PR where I extend snapshot tests to other OSes or scenarios, XenServer however is not one of them :)

ekohl reviewed (dismissed):
The concept of external tests deserves a highlight on Discourse.

lzap commented:
Rebased.

ezr-ondrej reviewed (dismissed):
Great addition! We need one more rebase though

lzap commented:
And rebased.

ezr-ondrej commented:
[test foreman]

ezr-ondrej commented:
Finally figured what is happening with the test failures, you've just missed one snapshot xD
Fixed, rebased and gonna merge if test :green_apple:

ezr-ondrej commented:
[test katello]

ezr-ondrej reviewed (approved):
More testing, better stabilization.
Thanks @lzap :+1:  and sorry for the delay :disappointed:
